aten/src/ATen/BatchedTensorImpl.cpp

```diff
@@ -101,6 +101,35 @@ bool BatchedTensorImpl::has_storage() const {
 }
 #endif
 
+void BatchedTensorImpl::replace_(const TensorImpl* other_impl) {
```

Comment:
Do we really want to support replace_ on BatchedTensorImpl? Isn't in functorch we supposed to have a FunctionalizeTensor that handles the replacing?

Reply:
Let me know if you think I'm missing something, but the reason I added it was to maintain the semantics for programs that mutate their inputs (although I dunno what the state of `BatchedTensorImpl` in core vs. the copy in functorch is. For now I was just trying to keep them in sync in my PR). E.g.
```
def f(x):
  y = torch.ones(2)
  x.add_(y)
  return x

batched_input = torch.ones(2, 2)
vmap(functionalize(f))(batched_input)
# batched_input should now be mutated: [[2, 2], [2, 2]]
```

Functionalizing the program eventually turned that inplace call into `tmp = torch.add(x, y); x.replace_(tmp)`. To maintain the fact that the input got mutated, we need the original `batched_inputs` to now point to the storage of `tmp`.

In this particular case, the types that we have are:
```
batched_inputs = [a] TensorImpl([b] storage)
(inside of f) x = FunctionalTensorWrapper(BatchedTensorImpl([a] TensorImpl([b] storage)))
(inside of f) y = FunctionalTensorWrapper(TensorImpl(storage))
(inside of f) tmp = FunctionalTensorWrapper(BatchedTensorImpl([c] TensorImpl([d] storage)))
```

Where `[a, b, c, d]` refer to particular `TensorImpl/StorageImpl` references. We want the `TensorImpl` from the input (`a`) to point to the storage of the new value (`d`). Since you can have arbitrary nesting, I thought it'd be cleaner to implement that by telling each wrapper how to replace itself with another wrapper of the same type (until we hit the `TensorImpl` base case, where we swap storages).

Reply:
I can try to understand your argument in more detail tomorrow. But the quick heuristic I use is: what if the batched tensor is a view? If your thing doesn't propagate the mutation to the base tensor, it's wrong.